File: src/GitHub/RepoUrlParser.cs
using System.Text.RegularExpressions;

namespace GitHubWallpaper.GitHub;

/// <summary>
/// Разбирает строки вида <c>owner/repo</c> и <c>https://github.com/owner/repo</c>.
/// </summary>
internal static partial class RepoUrlParser
{
    private static readonly Regex NamePattern = ValidNameRegex();

    /// <summary>
    /// Пытается разобрать строку в <see cref="RepoReference"/>.
    /// </summary>
    public static bool TryParse(string? input, out RepoReference reference)
    {
        reference = default;

        if (string.IsNullOrWhiteSpace(input))
        {
            return false;
        }

        var trimmed = input.Trim();

        if (TryParseFromUri(trimmed, out reference))
        {
            return true;
        }

        if (TryParseFromSlug(trimmed, out reference))
        {
            return true;
        }

        return false;
    }

    /// <summary>
    /// Разбирает строку в <see cref="RepoReference"/> или выбрасывает <see cref="FormatException"/>.
    /// </summary>
    public static RepoReference Parse(string input)
    {
        if (TryParse(input, out var reference))
        {
            return reference;
        }

        throw new FormatException(
            "Ожидается формат owner/repo или https://github.com/owner/repo.");
    }

    private static bool TryParseFromUri(string input, out RepoReference reference)
    {
        reference = default;

        var candidate = input.Contains("://", StringComparison.Ordinal)
            ? input
            : input.StartsWith("github.com/", StringComparison.OrdinalIgnoreCase)
                || input.StartsWith("www.github.com/", StringComparison.OrdinalIgnoreCase)
                ? $"https://{input.TrimStart('/')}"
                : null;

        if (candidate is null || !Uri.TryCreate(candidate, UriKind.Absolute, out var uri))
        {
            return false;
        }

        if (!IsGitHubHost(uri.Host))
        {
            return false;
        }

        return TryCreateFromPath(uri.AbsolutePath, out reference);
    }

    private static bool TryParseFromSlug(string input, out RepoReference reference)
    {
        reference = default;

        var path = input;

        var hashIndex = path.IndexOf('#');
        if (hashIndex >= 0)
        {
            path = path[..hashIndex];
        }

        var queryIndex = path.IndexOf('?');
        if (queryIndex >= 0)
        {
            path = path[..queryIndex];
        }

        return TryCreateFromPath(path, out reference);
    }

    private static bool TryCreateFromPath(string path, out RepoReference reference)
    {
        reference = default;

        var segments = path
            .Split('/', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        if (segments.Length < 2)
        {
            return false;
        }

        var owner = segments[0];
        var repo = segments[1];

        if (repo.EndsWith(".git", StringComparison.OrdinalIgnoreCase))
        {
            repo = repo[..^4];
        }

        if (!IsValidName(owner) || !IsValidName(repo))
        {
            return false;
        }

        reference = new RepoReference(owner, repo);
        return true;
    }

    private static bool IsGitHubHost(string host) =>
        host.Equals("github.com", StringComparison.OrdinalIgnoreCase)
        || host.Equals("www.github.com", StringComparison.OrdinalIgnoreCase);

    private static bool IsValidName(string name) =>
        !string.IsNullOrEmpty(name) && NamePattern.IsMatch(name);

    [GeneratedRegex("^[A-Za-z0-9][A-Za-z0-9._-]*$")]
    private static partial Regex ValidNameRegex();
}

File: python/github_wallpaper/repo_url_parser.py
"""Разбор строк owner/repo и URL GitHub (минимальный порт RepoUrlParser.cs)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from urllib.parse import urlparse

_VALID_NAME = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


@dataclass(frozen=True, slots=True)
class RepoReference:
    owner: str
    repo: str


def try_parse(input_value: str | None) -> RepoReference | None:
    """Разбирает owner/repo или https://github.com/owner/repo."""
    if not input_value or not input_value.strip():
        return None

    trimmed = input_value.strip()
    reference = _try_parse_from_uri(trimmed)
    if reference is not None:
        return reference

    return _try_parse_from_slug(trimmed)


def _try_parse_from_uri(input_value: str) -> RepoReference | None:
    candidate: str | None
    if "://" in input_value:
        candidate = input_value
    elif input_value.lower().startswith(("github.com/", "www.github.com/")):
        candidate = f"https://{input_value.lstrip('/')}"
    else:
        candidate = None

    if candidate is None:
        return None

    parsed = urlparse(candidate)
    if parsed.scheme not in ("http", "https"):
        return None

    host = parsed.hostname or ""
    if host.lower() not in ("github.com", "www.github.com"):
        return None

    return _try_create_from_path(parsed.path)


def _try_parse_from_slug(input_value: str) -> RepoReference | None:
    path = input_value.split("#", 1)[0].split("?", 1)[0]
    return _try_create_from_path(path)


def _try_create_from_path(path: str) -> RepoReference | None:
    segments = [segment.strip() for segment in path.split("/") if segment.strip()]
    if len(segments) < 2:
        return None

    owner = segments[0]
    repo = segments[1]
    if repo.lower().endswith(".git"):
        repo = repo[:-4]

    if not _is_valid_name(owner) or not _is_valid_name(repo):
        return None

    return RepoReference(owner=owner, repo=repo)


def _is_valid_name(name: str) -> bool:
    return bool(name) and _VALID_NAME.match(name) is not None
